content/flexible-varargs.md

```diff
@@ -0,0 +1,350 @@
+
+// Desugaring
+val total = sum((IArray(0) ++ numbers1 ++ numbers2 ++ IArray(4))*) // 10
+```
+
+The implementation for patterns could be something like
+
+```scala
+// User Code
+numbers match {
+  case Seq(1, numbers2*, 3) => println(numbers2) // Seq(2)
+}
+
+// Desugaring Helper
+class VarargsMatchHelper(beforeCount: Int, afterCount: Int) {
+  def unapply[T](values: Seq[T]): Option[(Seq[T], Seq[T], Seq[T])] = {
+    Option.when (values.length >= beforeCount + afterCount){
+      (
+        values.take(beforeCount),
+        values.drop(beforeCount).dropRight(afterCount), 
+        values.takeRight(afterCount)
+      )
+    }
+  }
+}
+
+// Desugaring Helper
+val VarargsMatcher = new VarargsMatchHelper(1, 1)
+numbers match {
+  case VarargsMatcher(Seq(1), numbers2, Seq(3)) => println(numbers2) // Seq(2)
+}
+```
+
+## Limitations
+
+
+### Single `*` in pattern matching
+
+One major limitation is that while expressions support unpacking multiple `*`s in a varargs,
+pattern matching can only support a single `*`. That is because a varargs pattern with
+multiple `*` sub-patterns may have multiple possible ways of assigning the individual
+elements to each sub-pattern, and depending on the sub-patterns not all such assignments
+may be valid. Thus there is no way to implement it efficiently in general, as it would 
+require an expensive (`O(2^n)`) backtracking search to try and find a valid assignment 
+of the elements that satisfies all sub-patterns. 
+
+### No Specific Performance Optimizations
+
+As proposed, the desugaring just relies on `IArray()` and `++` to construct the final 
+sequence that will be passed to varargs. We don't want to use `Seq` because it returns
+a `List`, and `List#++` is very inefficient. But we do not do any further optimizations,
+and anyone who hits performance issues with flexible vararg unpacking can always rewrite
+it themselves manually constructing an `mutable.ArrayBuffer` or `mutable.ArrayDeque`.
+
+### Not supporting unpacking arbitrary `Iterable`s
+
+Given we propose to allow unpacking `Seq` and `Option`, its worth asking if we should 
+support unpacking `Iterable` or `IterableOnce`. This proposal avoids those types for now,
+but if anyone wants to unpack them it's trivial to call `.toSeq`. For `Option`, we think
+that the use case of calling a vararg method with optional values is frequent enough that 
+it deserves special support, whereas calling a vararg method with a `Set` or a `Map`
+doesn't happen enough to be worth special casing.
+
+## Alternatives
+
+Apart from the manual workflows described above, one alternative is to use implicit conversions
+to a target type, i.e. the "magnet pattern". For example:
+
+```scala
+case class Summable(value: Seq[Int])
+implicit def seqSummable(value: Seq[Int]) = Summable(value) 
+implicit def singleSummable(value: Int) = Summable(Seq(value))
+implicit def optionSummable(value: Option[Int]) = Summable(value.toSeq)
+
+def sum(x: Int*) = x.sum
```

Comment:
I suspect this was supposed to be
```suggestion
def sum(x: Summable*) = x.sum
```

to trigger the implicit conversions.

Reply:
Thanks!